clang/lib/CIR/Dialect/Transforms/FlattenCFG.cpp

```diff
@@ -906,6 +911,72 @@ class CIRTernaryOpFlattening : public mlir::OpRewritePattern<cir::TernaryOp> {
   }
 };
 
+struct UnifyReturn : public OpRewritePattern<ReturnOp> {
+  using OpRewritePattern<ReturnOp>::OpRewritePattern;
+
+  UnifyReturn(MLIRContext *context, cir::FuncOp func, Block *retBlock)
+      : OpRewritePattern<ReturnOp>(context), func(func), retBlock(retBlock) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(cir::ReturnOp ret,
+                  mlir::PatternRewriter &rewriter) const override {
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+    auto fn = ret->getParentOfType<cir::FuncOp>();
+    if (!fn || fn != func)
+      return mlir::failure();
+    // Replace 'return' with 'br <retBlock>'
+    rewriter.replaceOpWithNewOp<cir::BrOp>(ret, ret.getInput(), retBlock);
+    return mlir::success();
+  }
+
+private:
+  cir::FuncOp func; // The parent funcOp
+  Block *retBlock;  // The dedicated return block
+};
+
+void FlattenCFGPass::unifyReturn(cir::FuncOp func) {
+  if (func.getRegion().empty())
+    return;
+
+  // Collect operations to apply patterns.
+  llvm::SmallVector<Operation *, 16> ops;
```

Comment:
It is unclear to me why you have this type erasure and why you cannot keep `cir::ReturnOp` here, such as:
```suggestion
  llvm::SmallVector<cir::ReturnOp, 16> returnOps;
```
OK, I reply to myself, because this is required later by `applyOpPatternsGreedily()`. :smile:
But probably renaming `ops` to `returnOps` would be clearer.